src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java

```diff
@@ -166,10 +234,198 @@ private boolean validateNucleusConfig(CompletableFuture<DeploymentResult> totall
                         .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE, e));
                 return false;
             }
+
+            long configTimeout = Duration.ofSeconds(DEFAULT_TIMEOUT_SECOND).toMillis();
+            if (timeoutSec != null) {
+                configTimeout = Duration.ofSeconds(timeoutSec).toMillis();
+            }
+
+            if (configTimeout == 0
+                    || !deviceConfiguration.isDeviceConfiguredToTalkToCloud()) {
+                logger.atDebug().log("Skipping connectivity validation");
+                return true;
+            }
+            try {
+                // Check that MQTT client has reconnected
+                logger.atDebug().log("Checking MQTT Reconnected");
+                // TODO - use ScheduledExecuterService for wait and retry
+                // TODO - make a separate function to create and wait
+                MqttClientConnection connection = createMqttConnection(nucleusConfig);
+                connection.connect().get();
+                // Close after
+                connection.close();
+
+                // Check that HTTP client works
+                logger.atDebug().log("Checking HTTP Reconnected");
+                // TODO - use ScheduledExecuterService for wait and retry
+                // TODO - make a separate function to create and wait
+                GreengrassV2DataClient greengrassV2DataClient = createGGv2DataClient(nucleusConfig);
+                // move code out of thing Group helper
+                thingGroupHelper.waitForReconnect(configTimeout, greengrassV2DataClient);
+
+            } catch (MqttException | GeneralSecurityException | IOException
+                     | InterruptedException | ExecutionException
+                     | TLSAuthException | InvalidEnvironmentStageException
+                     | ComponentConfigurationValidationException e) {
+                logger.atError().cause(e).log("Nucleus connectivity validation failed");
+                totallyCompleteFuture
+                        .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE, e));
+                return false;
+            }
         }
         return true;
     }
 
+    private MqttClientConnection createMqttConnection(Map<String, Object> nucleusConfig) {
+        AwsIotMqttConnectionBuilder builder;
+        try {
+            builder = securityService.getDeviceIdentityMqttConnectionBuilder();
+        } catch (MqttConnectionProviderException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While re-throwing the caught exception with modifications , information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

Suggested remediation:
Consider passing the exception object either as  &#34;rootCause&#34; or as an inner exception parameter, to the constructor of the new exception before throwing new exception. This will help keep stack trace.
```
@@ -283,3 +283,3 @@
         } catch (MqttConnectionProviderException e) {
-            throw new MqttException(e.getMessage());
+            throw new MqttException(e.getMessage(), e);
         }

```